gsplat/cuda/_wrapper.py

```diff
@@ -0,0 +1,1120 @@
+        compute_covar: Whether to compute covariance matrices. Default: True. If False,
+            the returned covariance matrices will be None.
+        compute_preci: Whether to compute precision matrices. Default: True. If False,
+            the returned precision matrices will be None.
+        triu: If True, the return matrices will be upper triangular. Default: False.
+
+    Returns:
+        A tuple:
+
+        - **Covariance matrices**. If `triu` is True the returned shape is [N, 6], otherwise [N, 3, 3].
+        - **Precision matrices**. If `triu` is True the returned shape is [N, 6], otherwise [N, 3, 3].
+    """
+    assert quats.dim() == 2 and quats.size(1) == 4, quats.size()
+    assert scales.dim() == 2 and scales.size(1) == 3, scales.size()
+    quats = quats.contiguous()
+    scales = scales.contiguous()
+    covars, precis = _QuatScaleToCovarpreci.apply(
+        quats, scales, compute_covar, compute_preci, triu
+    )
+    return covars if compute_covar else None, precis if compute_preci else None
+
+
+def persp_proj(
+    means: Tensor,  # [C, N, 3]
+    covars: Tensor,  # [C, N, 3, 3]
+    Ks: Tensor,  # [C, 3, 3]
+    width: int,
+    height: int,
+) -> Tuple[Tensor, Tensor]:
+    """Perspective projection on Gaussians.
+
+    Args:
+        means: Gaussian means. [C, N, 3]
+        covars: Gaussian covariances. [C, N, 3, 3]
+        Ks: Camera intrinsics. [C, 3, 3]
+        width: Image width.
+        height: Image height.
+
+    Returns:
+        A tuple:
+
+        - **Projected means**. [C, N, 2]
+        - **Projected covariances**. [C, N, 2, 2]
+    """
+    C, N, _ = means.shape
+    assert means.shape == (C, N, 3), means.size()
+    assert covars.shape == (C, N, 3, 3), covars.size()
+    assert Ks.shape == (C, 3, 3), Ks.size()
+    means = means.contiguous()
+    covars = covars.contiguous()
+    Ks = Ks.contiguous()
+    return _PerspProj.apply(means, covars, Ks, width, height)
+
+
+def world_to_cam(
+    means: Tensor,  # [N, 3]
+    covars: Tensor,  # [N, 3, 3]
+    viewmats: Tensor,  # [C, 4, 4]
+) -> Tuple[Tensor, Tensor]:
+    """Transforms Gaussians from world to camera coordinate system.
+
+    Args:
+        means: Gaussian means. [N, 3]
+        covars: Gaussian covariances. [N, 3, 3]
+        viewmats: World-to-camera transformation matrices. [C, 4, 4]
+
+    Returns:
+        A tuple:
+
+        - **Gaussian means in camera coordinate system**. [C, N, 3]
+        - **Gaussian covariances in camera coordinate system**. [C, N, 3, 3]
+    """
+    C = viewmats.size(0)
+    N = means.size(0)
+    assert means.size() == (N, 3), means.size()
+    assert covars.size() == (N, 3, 3), covars.size()
+    assert viewmats.size() == (C, 4, 4), viewmats.size()
+    means = means.contiguous()
+    covars = covars.contiguous()
+    viewmats = viewmats.contiguous()
+    return _WorldToCam.apply(means, covars, viewmats)
+
+
+def projection(
+    means: Tensor,  # [N, 3]
+    covars: Tensor,  # [N, 6] or None
```

Comment:
(noted this when I was porting https://github.com/nerfstudio-project/gsplat/compare/cuda_v2...hturki:gsplat:ht/cuda_v2_batch?expand=1 to this branch)

I would suggest annotating the parameters that might be None as Optional[Tensor]

Reply:
oh yes good catch! I think was generally following this rule but I clearly missed this one!